File: net/crates/net/sdk-ts/test/groups.test.ts
// Stage 2 tests for the groups surface — ReplicaGroup, ForkGroup,
// StandbyGroup. Mirrors the Rust SDK's `groups_surface.rs`.

import { afterEach, describe, expect, it } from 'vitest';

import {
  DaemonError,
  DaemonRuntime,
  ForkGroup,
  GroupError,
  Identity,
  MeshNode,
  ReplicaGroup,
  StandbyGroup,
  type MeshDaemon,
} from '../src';

const PSK = '42'.repeat(32);

// Start high to avoid collisions with other test files in the
// suite. Each port is used exactly once per vitest run.
let portSeed = 31_500;
function nextPort(): string {
  return `127.0.0.1:${portSeed++}`;
}

async function buildMesh(): Promise<MeshNode> {
  return MeshNode.create({ bindAddr: nextPort(), psk: PSK });
}

class NoopDaemon implements MeshDaemon {
  readonly name = 'noop';
  process(): Buffer[] {
    return [];
  }
}

/** Build a started runtime with the given number of synthetic
 *  peer entries in the capability index, so `place_with_spread`
 *  has enough candidates for ≥ 2 member groups. */
async function runtimeWithPeers(extraPeers: number): Promise<{ rt: DaemonRuntime; mesh: MeshNode }> {
  const mesh = await buildMesh();
  for (let i = 1; i <= extraPeers; i++) {
    // Synthetic node IDs above 0x1000_0000 so they never collide
    // with real node IDs derived from ed25519 pubkeys.
    mesh._testInjectSyntheticPeer(BigInt(0x1000_0000_0000_0000) + BigInt(i));
  }
  const rt = DaemonRuntime.create(mesh);
  rt.registerFactory('noop', () => new NoopDaemon());
  await rt.start();
  return { rt, mesh };
}

const seed = (byte: number): Buffer => Buffer.alloc(32, byte);

describe('ReplicaGroup', () => {
  const cleanups: Array<() => Promise<void>> = [];
  afterEach(async () => {
    while (cleanups.length > 0) {
      const fn = cleanups.pop();
      if (fn) {
        try {
          await fn();
        } catch {
          // Best-effort teardown
        }
      }
    }
  });

  it('spawn registers all members and reports healthy aggregate', async () => {
    const { rt, mesh } = await runtimeWithPeers(3);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    const group = await ReplicaGroup.spawn(rt, 'noop', {
      replicaCount: 3,
      groupSeed: seed(0x11),
      lbStrategy: 'round-robin',
    });

    expect(group.replicaCount).toBe(3);
    expect(group.healthyCount).toBe(3);
    expect(group.health.status).toBe('healthy');
    expect(rt.daemonCount()).toBe(3);
    expect(group.replicas).toHaveLength(3);
    for (const r of group.replicas) {
      expect(r.healthy).toBe(true);
      expect(r.originHash).not.toBe(0n);
    }
  });

  it('routeEvent returns a live member origin', async () => {
    const { rt, mesh } = await runtimeWithPeers(3);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    const group = await ReplicaGroup.spawn(rt, 'noop', {
      replicaCount: 3,
      groupSeed: seed(0x22),
      lbStrategy: 'consistent-hash',
    });

    const live = new Set(group.replicas.map((m) => m.originHash));
    for (let i = 0; i < 30; i++) {
      const origin = group.routeEvent({ routingKey: `req-${i}` });
      expect(live.has(origin)).toBe(true);
    }
  });

  it('scaleTo grows and shrinks the daemon registry', async () => {
    const { rt, mesh } = await runtimeWithPeers(5);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    const group = await ReplicaGroup.spawn(rt, 'noop', {
      replicaCount: 2,
      groupSeed: seed(0x33),
      lbStrategy: 'round-robin',
    });
    expect(rt.daemonCount()).toBe(2);

    await group.scaleTo(5);
    expect(group.replicaCount).toBe(5);
    expect(rt.daemonCount()).toBe(5);

    await group.scaleTo(1);
    expect(group.replicaCount).toBe(1);
    expect(rt.daemonCount()).toBe(1);
  });

  it('spawn before runtime start throws GroupError kind not-ready', async () => {
    const mesh = await buildMesh();
    cleanups.push(() => mesh.shutdown());
    const rt = DaemonRuntime.create(mesh);
    cleanups.push(() => rt.shutdown());
    rt.registerFactory('noop', () => new NoopDaemon());
    // Intentionally skip rt.start()

    try {
      await ReplicaGroup.spawn(rt, 'noop', {
        replicaCount: 2,
        groupSeed: seed(0x44),
        lbStrategy: 'round-robin',
      });
      expect.fail('expected GroupError');
    } catch (e) {
      expect(e).toBeInstanceOf(GroupError);
      expect(e).toBeInstanceOf(DaemonError);
      expect((e as GroupError).kind).toBe('not-ready');
    }
  });

  it('spawn with unregistered kind throws factory-not-found', async () => {
    const { rt, mesh } = await runtimeWithPeers(2);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    try {
      await ReplicaGroup.spawn(rt, 'never-registered', {
        replicaCount: 2,
        groupSeed: seed(0x55),
        lbStrategy: 'round-robin',
      });
      expect.fail('expected GroupError');
    } catch (e) {
      expect(e).toBeInstanceOf(GroupError);
      expect((e as GroupError).kind).toBe('factory-not-found');
      expect((e as GroupError).requestedKind).toBe('never-registered');
    }
  });
});

describe('ForkGroup', () => {
  const cleanups: Array<() => Promise<void>> = [];
  afterEach(async () => {
    while (cleanups.length > 0) {
      const fn = cleanups.pop();
      if (fn) {
        try {
          await fn();
        } catch {
          // best-effort
        }
      }
    }
  });

  it('forks have unique origins and verifiable lineage', async () => {
    const { rt, mesh } = await runtimeWithPeers(4);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    const group = await ForkGroup.fork(rt, 'noop', 0xabcd_ef01n, 42n, {
      forkCount: 3,
      lbStrategy: 'round-robin',
    });

    expect(group.forkCount).toBe(3);
    expect(group.parentOrigin).toBe(0xabcd_ef01n);
    expect(group.forkSeq).toBe(42n);
    expect(group.verifyLineage()).toBe(true);

    const origins = new Set(group.members.map((m) => m.originHash));
    expect(origins.size).toBe(3);
    expect(group.forkRecords).toHaveLength(3);
  });

  it('fork before runtime start throws not-ready', async () => {
    const mesh = await buildMesh();
    cleanups.push(() => mesh.shutdown());
    const rt = DaemonRuntime.create(mesh);
    cleanups.push(() => rt.shutdown());
    rt.registerFactory('noop', () => new NoopDaemon());

    try {
      await ForkGroup.fork(rt, 'noop', 0x1234n, 1n, {
        forkCount: 2,
        lbStrategy: 'round-robin',
      });
      expect.fail('expected GroupError');
    } catch (e) {
      expect((e as GroupError).kind).toBe('not-ready');
    }
  });
});

describe('StandbyGroup', () => {
  const cleanups: Array<() => Promise<void>> = [];
  afterEach(async () => {
    while (cleanups.length > 0) {
      const fn = cleanups.pop();
      if (fn) {
        try {
          await fn();
        } catch {
          // best-effort
        }
      }
    }
  });

  it('spawn makes member 0 active and leaves standbys buffered-empty', async () => {
    const { rt, mesh } = await runtimeWithPeers(3);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    const group = await StandbyGroup.spawn(rt, 'noop', {
      memberCount: 3,
      groupSeed: seed(0x77),
    });

    expect(group.memberCount).toBe(3);
    expect(group.standbyCount).toBe(2);
    expect(group.activeIndex).toBe(0);
    expect(group.activeHealthy).toBe(true);
    expect(group.activeOrigin).not.toBe(0n);
    expect(group.bufferedEventCount).toBe(0);
    expect(group.memberRole(0)).toBe('active');
    expect(group.memberRole(1)).toBe('standby');
    expect(group.memberRole(99)).toBeNull();
  });

  it('member_count below 2 surfaces as invalid-config', async () => {
    const { rt, mesh } = await runtimeWithPeers(3);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    try {
      await StandbyGroup.spawn(rt, 'noop', {
        memberCount: 1,
        groupSeed: seed(0x88),
      });
      expect.fail('expected GroupError');
    } catch (e) {
      expect(e).toBeInstanceOf(GroupError);
      expect((e as GroupError).kind).toBe('invalid-config');
    }
  });

  it('spawn with unregistered kind throws factory-not-found', async () => {
    const { rt, mesh } = await runtimeWithPeers(3);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    try {
      await StandbyGroup.spawn(rt, 'never-registered', {
        memberCount: 2,
        groupSeed: seed(0x99),
      });
      expect.fail('expected GroupError');
    } catch (e) {
      expect((e as GroupError).kind).toBe('factory-not-found');
    }
  });

  it('rejects a groupSeed buffer that is not exactly 32 bytes', async () => {
    const { rt, mesh } = await runtimeWithPeers(3);
    cleanups.push(() => mesh.shutdown());
    cleanups.push(() => rt.shutdown());

    // Invalid: 16 bytes instead of 32. NAPI-side `parse_seed`
    // surfaces this as `daemon: group: invalid-config: ...`.
    try {
      await StandbyGroup.spawn(rt, 'noop', {
        memberCount: 2,
        groupSeed: Buffer.alloc(16, 0xaa),
      });
      expect.fail('expected GroupError');
    } catch (e) {
      expect((e as GroupError).kind).toBe('invalid-config');
    }
  });
});

File: go/groups_test.go
// Tests for the groups surface — Stage 4 of SDK_GROUPS_SURFACE_PLAN.md.
//
// Requires the `test_helpers` build tag so
// `TestInjectSyntheticPeer` (from `groups_testhelpers.go`) is
// compiled in. Run with:
//
//   DYLD_LIBRARY_PATH=.../target/release go test -tags test_helpers ./net/...
//
// and the cdylib must be built with `--features test-helpers`.

//go:build test_helpers

package net

import (
	"errors"
	"strings"
	"testing"
)

// runtimeWithPeers builds a started DaemonRuntime with a factory
// registered under "noop" and `extraPeers` synthetic capability
// entries in the mesh's index so `place_with_spread` has enough
// placement candidates for multi-member groups.
func runtimeWithPeers(t *testing.T, extraPeers int) (*DaemonRuntime, *MeshNode) {
	t.Helper()
	mesh := newLocalMesh(t)
	for i := 1; i <= extraPeers; i++ {
		// Synthetic node IDs above 0x1000_0000 so they never
		// collide with real node IDs derived from ed25519 pubkeys.
		TestInjectSyntheticPeer(mesh, uint64(0x1000_0000_0000_0000)+uint64(i))
	}
	rt, err := NewDaemonRuntime(mesh)
	if err != nil {
		t.Fatalf("NewDaemonRuntime: %v", err)
	}
	if err := rt.RegisterFactoryFunc("noop", func() MeshDaemon {
		return &noopDaemon{}
	}); err != nil {
		t.Fatalf("RegisterFactoryFunc: %v", err)
	}
	if err := rt.Start(); err != nil {
		t.Fatalf("Start: %v", err)
	}
	return rt, mesh
}

type noopDaemon struct{}

func (noopDaemon) Process(CausalEvent) ([][]byte, error) { return nil, nil }

func seed(b byte) []byte {
	s := make([]byte, 32)
	for i := range s {
		s[i] = b
	}
	return s
}

// ----------------------------------------------------------------
// ReplicaGroup
// ----------------------------------------------------------------

func TestReplicaGroup_SpawnRegistersMembersAndReportsHealthy(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 3)
	defer mesh.Shutdown()
	defer rt.Close()

	g, err := NewReplicaGroup(rt, "noop", ReplicaGroupConfig{
		ReplicaCount: 3,
		GroupSeed:    seed(0x11),
		LBStrategy:   StrategyRoundRobin,
	})
	if err != nil {
		t.Fatalf("NewReplicaGroup: %v", err)
	}
	defer g.Close()

	if n := g.ReplicaCount(); n != 3 {
		t.Errorf("ReplicaCount = %d, want 3", n)
	}
	if n := g.HealthyCount(); n != 3 {
		t.Errorf("HealthyCount = %d, want 3", n)
	}
	if h := g.Health(); h.Status != "healthy" {
		t.Errorf("Health.Status = %q, want healthy", h.Status)
	}
	if n := rt.DaemonCount(); n != 3 {
		t.Errorf("rt.DaemonCount() = %d, want 3", n)
	}
	members := g.Replicas()
	if len(members) != 3 {
		t.Fatalf("len(Replicas()) = %d, want 3", len(members))
	}
	for _, m := range members {
		if !m.Healthy {
			t.Errorf("member %d not healthy", m.Index)
		}
		if m.OriginHash == 0 {
			t.Errorf("member %d has zero origin_hash", m.Index)
		}
	}
}

func TestReplicaGroup_RouteEventReturnsLiveMemberOrigin(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 3)
	defer mesh.Shutdown()
	defer rt.Close()

	g, err := NewReplicaGroup(rt, "noop", ReplicaGroupConfig{
		ReplicaCount: 3,
		GroupSeed:    seed(0x22),
		LBStrategy:   StrategyConsistentHash,
	})
	if err != nil {
		t.Fatalf("NewReplicaGroup: %v", err)
	}
	defer g.Close()

	live := make(map[uint64]bool)
	for _, m := range g.Replicas() {
		live[m.OriginHash] = true
	}
	for i := 0; i < 30; i++ {
		origin, err := g.RouteEvent("req-" + string(rune('a'+i%26)))
		if err != nil {
			t.Fatalf("RouteEvent iter %d: %v", i, err)
		}
		if !live[origin] {
			t.Errorf("RouteEvent iter %d returned unknown origin %#x", i, origin)
		}
	}
}

func TestReplicaGroup_ScaleUpAndDown(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 5)
	defer mesh.Shutdown()
	defer rt.Close()

	g, err := NewReplicaGroup(rt, "noop", ReplicaGroupConfig{
		ReplicaCount: 2,
		GroupSeed:    seed(0x33),
		LBStrategy:   StrategyRoundRobin,
	})
	if err != nil {
		t.Fatalf("NewReplicaGroup: %v", err)
	}
	defer g.Close()

	if err := g.ScaleTo(5); err != nil {
		t.Fatalf("ScaleTo(5): %v", err)
	}
	if g.ReplicaCount() != 5 {
		t.Errorf("ReplicaCount after scale-up = %d", g.ReplicaCount())
	}
	if rt.DaemonCount() != 5 {
		t.Errorf("rt.DaemonCount = %d after scale-up", rt.DaemonCount())
	}

	if err := g.ScaleTo(1); err != nil {
		t.Fatalf("ScaleTo(1): %v", err)
	}
	if g.ReplicaCount() != 1 {
		t.Errorf("ReplicaCount after scale-down = %d", g.ReplicaCount())
	}
	if rt.DaemonCount() != 1 {
		t.Errorf("rt.DaemonCount = %d after scale-down", rt.DaemonCount())
	}
}

func TestReplicaGroup_SpawnBeforeStartErrorsNotReady(t *testing.T) {
	mesh := newLocalMesh(t)
	defer mesh.Shutdown()
	rt, err := NewDaemonRuntime(mesh)
	if err != nil {
		t.Fatalf("NewDaemonRuntime: %v", err)
	}
	defer rt.Close()
	if err := rt.RegisterFactoryFunc("noop", func() MeshDaemon {
		return &noopDaemon{}
	}); err != nil {
		t.Fatalf("RegisterFactoryFunc: %v", err)
	}
	// Intentionally skip rt.Start()

	_, err = NewReplicaGroup(rt, "noop", ReplicaGroupConfig{
		ReplicaCount: 2,
		GroupSeed:    seed(0x44),
		LBStrategy:   StrategyRoundRobin,
	})
	if err == nil {
		t.Fatal("expected not-ready error")
	}
	var ge *GroupError
	if !errors.As(err, &ge) {
		t.Fatalf("err = %T %v, want *GroupError", err, err)
	}
	if ge.Kind != GroupErrNotReady {
		t.Errorf("Kind = %q, want %q", ge.Kind, GroupErrNotReady)
	}
}

func TestReplicaGroup_SpawnUnknownKindErrorsFactoryNotFound(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 2)
	defer mesh.Shutdown()
	defer rt.Close()

	_, err := NewReplicaGroup(rt, "never-registered", ReplicaGroupConfig{
		ReplicaCount: 2,
		GroupSeed:    seed(0x55),
		LBStrategy:   StrategyRoundRobin,
	})
	if err == nil {
		t.Fatal("expected factory-not-found error")
	}
	var ge *GroupError
	if !errors.As(err, &ge) {
		t.Fatalf("err = %T %v, want *GroupError", err, err)
	}
	if ge.Kind != GroupErrFactoryNotFound {
		t.Errorf("Kind = %q, want %q", ge.Kind, GroupErrFactoryNotFound)
	}
}

func TestReplicaGroup_InvalidSeedErrors(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 2)
	defer mesh.Shutdown()
	defer rt.Close()

	_, err := NewReplicaGroup(rt, "noop", ReplicaGroupConfig{
		ReplicaCount: 2,
		GroupSeed:    []byte("short"),
		LBStrategy:   StrategyRoundRobin,
	})
	if err == nil {
		t.Fatal("expected invalid-seed error")
	}
	var ge *GroupError
	if !errors.As(err, &ge) {
		t.Fatalf("err = %T %v, want *GroupError", err, err)
	}
	if ge.Kind != GroupErrInvalidConfig {
		t.Errorf("Kind = %q, want invalid-config", ge.Kind)
	}
	if !strings.Contains(ge.Error(), "group_seed must be 32 bytes") {
		t.Errorf("error message = %q, missing group_seed hint", ge.Error())
	}
}

// ----------------------------------------------------------------
// ForkGroup
// ----------------------------------------------------------------

func TestForkGroup_UniqueOriginsAndVerifiableLineage(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 4)
	defer mesh.Shutdown()
	defer rt.Close()

	g, err := NewForkGroup(rt, "noop", 0xABCDEF01, 42, ForkGroupConfig{
		ForkCount:  3,
		LBStrategy: StrategyRoundRobin,
	})
	if err != nil {
		t.Fatalf("NewForkGroup: %v", err)
	}
	defer g.Close()

	if g.ForkCount() != 3 {
		t.Errorf("ForkCount = %d, want 3", g.ForkCount())
	}
	if g.ParentOrigin() != 0xABCDEF01 {
		t.Errorf("ParentOrigin = %#x, want 0xABCDEF01", g.ParentOrigin())
	}
	if g.ForkSeq() != 42 {
		t.Errorf("ForkSeq = %d, want 42", g.ForkSeq())
	}
	if !g.VerifyLineage() {
		t.Error("VerifyLineage() = false, want true")
	}

	origins := make(map[uint64]bool)
	for _, m := range g.Members() {
		origins[m.OriginHash] = true
	}
	if len(origins) != 3 {
		t.Errorf("unique origins = %d, want 3", len(origins))
	}
	if records := g.ForkRecords(); len(records) != 3 {
		t.Errorf("len(ForkRecords) = %d, want 3", len(records))
	}
}

func TestForkGroup_ForkBeforeStartErrorsNotReady(t *testing.T) {
	mesh := newLocalMesh(t)
	defer mesh.Shutdown()
	rt, err := NewDaemonRuntime(mesh)
	if err != nil {
		t.Fatalf("NewDaemonRuntime: %v", err)
	}
	defer rt.Close()
	_ = rt.RegisterFactoryFunc("noop", func() MeshDaemon { return &noopDaemon{} })
	// No Start()

	_, err = NewForkGroup(rt, "noop", 0x1234, 1, ForkGroupConfig{
		ForkCount:  2,
		LBStrategy: StrategyRoundRobin,
	})
	if err == nil {
		t.Fatal("expected not-ready error")
	}
	var ge *GroupError
	if !errors.As(err, &ge) {
		t.Fatalf("err = %T %v", err, err)
	}
	if ge.Kind != GroupErrNotReady {
		t.Errorf("Kind = %q, want not-ready", ge.Kind)
	}
}

// ----------------------------------------------------------------
// StandbyGroup
// ----------------------------------------------------------------

func TestStandbyGroup_MemberZeroIsActive(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 3)
	defer mesh.Shutdown()
	defer rt.Close()

	g, err := NewStandbyGroup(rt, "noop", StandbyGroupConfig{
		MemberCount: 3,
		GroupSeed:   seed(0x77),
	})
	if err != nil {
		t.Fatalf("NewStandbyGroup: %v", err)
	}
	defer g.Close()

	if g.MemberCount() != 3 {
		t.Errorf("MemberCount = %d, want 3", g.MemberCount())
	}
	if g.StandbyCount() != 2 {
		t.Errorf("StandbyCount = %d, want 2", g.StandbyCount())
	}
	if g.ActiveIndex() != 0 {
		t.Errorf("ActiveIndex = %d, want 0", g.ActiveIndex())
	}
	if !g.ActiveHealthy() {
		t.Error("ActiveHealthy = false, want true")
	}
	if g.ActiveOrigin() == 0 {
		t.Error("ActiveOrigin = 0, want non-zero")
	}
	if g.BufferedEventCount() != 0 {
		t.Errorf("BufferedEventCount = %d, want 0", g.BufferedEventCount())
	}
	if r := g.MemberRole(0); r != "active" {
		t.Errorf("MemberRole(0) = %q, want active", r)
	}
	if r := g.MemberRole(1); r != "standby" {
		t.Errorf("MemberRole(1) = %q, want standby", r)
	}
	if r := g.MemberRole(99); r != "" {
		t.Errorf("MemberRole(99) = %q, want empty", r)
	}
}

func TestStandbyGroup_MemberCountBelowTwoRejected(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 3)
	defer mesh.Shutdown()
	defer rt.Close()

	_, err := NewStandbyGroup(rt, "noop", StandbyGroupConfig{
		MemberCount: 1,
		GroupSeed:   seed(0x88),
	})
	if err == nil {
		t.Fatal("expected invalid-config error")
	}
	var ge *GroupError
	if !errors.As(err, &ge) {
		t.Fatalf("err = %T %v", err, err)
	}
	if ge.Kind != GroupErrInvalidConfig {
		t.Errorf("Kind = %q, want invalid-config", ge.Kind)
	}
}

func TestStandbyGroup_UnknownKindErrorsFactoryNotFound(t *testing.T) {
	rt, mesh := runtimeWithPeers(t, 3)
	defer mesh.Shutdown()
	defer rt.Close()

	_, err := NewStandbyGroup(rt, "never-registered", StandbyGroupConfig{
		MemberCount: 2,
		GroupSeed:   seed(0x99),
	})
	if err == nil {
		t.Fatal("expected factory-not-found error")
	}
	var ge *GroupError
	if !errors.As(err, &ge) {
		t.Fatalf("err = %T %v", err, err)
	}
	if ge.Kind != GroupErrFactoryNotFound {
		t.Errorf("Kind = %q, want factory-not-found", ge.Kind)
	}
}
